sdk/azidentity/aad_identity_client.go

```diff
@@ -238,6 +300,120 @@ func (c *aadIdentityClient) createUsernamePasswordAuthRequest(tenantID string, c
 	return msg, nil
 }
 
+func createDeviceCodeResult(res *azcore.Response) (*DeviceCodeResult, error) {
+	value := &DeviceCodeResult{}
+	if err := json.Unmarshal(res.Payload, &value); err != nil {
+		return nil, fmt.Errorf("DeviceCodeResult: %w", err)
+	}
+	return value, nil
+}
+
+// authenticateDeviceCode creates a device code authentication request and returns an Access Token or
+// an error in case of authentication failure.
+// ctx: The current request context
+// tenantID: The Azure Active Directory tenant (directory) Id of the service principal
+// clientID: The client (application) ID of the service principal
+// deviceCode: The device code associated with the request
+// scopes: The scopes required for the token
+func (c *aadIdentityClient) authenticateDeviceCode(ctx context.Context, tenantID string, clientID string, deviceCode string, scopes []string) (*tokenResponse, error) {
+	msg, err := c.createDeviceCodeAuthRequest(tenantID, clientID, deviceCode, scopes)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.pipeline.Do(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	// This should never happen under normal conditions
+	if resp == nil {
+		return nil, &AuthenticationFailedError{msg: "Something unexpected happened with the request and received a nil response"}
+	}
+
+	if resp.HasStatusCode(successStatusCodes[:]...) {
+		return c.createAccessToken(resp)
+	}
+
+	return nil, &AuthenticationFailedError{inner: newAADAuthenticationFailedError(resp)}
+}
+
+func (c *aadIdentityClient) createDeviceCodeAuthRequest(tenantID string, clientID string, deviceCode string, scopes []string) (*azcore.Request, error) {
+	if len(tenantID) == 0 { // if the user did not pass in a tenantID then the default value is set
+		tenantID = "organizations"
```

Comment:
Does this need to be documented in the exported API that calls this method?  I'm not saying it has to, but it seems magical.  If this is the normal expectation then no need to doc IMO but I'll defer to @schaabs

Reply:
I think it should be documented regardless of whether it is normally expected, simply to add clarity for users who haven't delved into the specifics of the flow they are using. Thanks for pointing this out. I'll add this to the info for the exported API.